### ISSUE TITLE
Sometimes, server measures speed incorrectly

### ISSUE THREAD

**Issue by LysergikProductions** [gcurtiss/protocol3#10]
When the user is moving at for example 25 bps, the server can measure as much as **twice** that speed during lag spikes / temporarily frozen ticks or for some other unknown reasons.

In the video you can see 45.5 bps client-side measured as 52.1 bps server-side.

This happens often going in straight lines (zero change in y value) at 44 bps client-side. This can be seen in the image included.
![44is53 8](https://user-images.githubusercontent.com/61183022/111393310-df9cb800-868e-11eb-946b-be138c105a71.png)

At it's worst, the server will measure as much as twice the actual speed.

STR:
1) Ensure server is allowing 48 bps at the time of testing.
2) Move in a straight line between 25 and 45 bps.
3) Occasionally observe issue

Results:
Grace period timer is started.

Notes:
The user is not kicked because of the grace period preventing disconnecting the player wrongly.
Low priority.

https://user-images.githubusercontent.com/61183022/111393723-b0d31180-868f-11eb-9381-dc69189e8fcd.mp4

**Comment by gcurtiss** [comment]
Hi there. This is because the way that speed is measured is based on TPS, not java time. This means that if the server's tps is 10, your speed will be measured as twice as fast compared to what it actually is. This is mitigated by a specially tested formula, but is far from perfect. This issue will be mitigated in the future by migrating the speed limit system to be dependent on java time as opposed to spigot time.

**Comment by gcurtiss** [comment]
ultradutch is fixing this it'll probably be done in 3.3.0

**Comment by gcurtiss** [comment]
@ultra64cmy cant find a commit pertaining to this, is this fixed or being worked on?

**Comment by ultra64cmy** [comment]
Collecting my thoughts on this before posting a longer reply. tl;dr it's complicated

**Comment by ultra64cmy** [comment]
Measuring speed in Minecraft is easy in theory: every 20 ticks (1 sec) you compare current position to previous position, and the length is blocks per sec. This is accurate as long as TPS is 20.0.

What happens if you have a long tick? On the client side, you will keep moving 48 blocks per sec but the server has stalled. Time is not progressing on the server until the tick completes. Let's say this takes 1 sec. Your position in your client is now 48 blocks further from your previous position, and your tick-timed (i.e., not wallclock time) speed is now double the speed limit.

The remainder of the 20 ticks happen. This could be instant if the last of the 20 ticks was long, or it could be up to 1 more second if it was the first tick that was long. Now consider what happens if you have 2+ long ticks.

We're currently dividing distance by wallclock time for speed, but even that isn't accurate. The 20 tick interrupt often fires at 0.95 sec at 20 tps, so what do you do there? Round up in increments of 50 ms (1 tick length)?

The most accurate way to measure speed is to do so every tick using wallclock time. I can implement that, but I suspect it would be too expensive under ~20 player load.

In conclusion, speed limit as it's currently implemented is a good 80% solution. I need ideas if we're gonna get the last 20% done.

**Comment by gcurtiss** [comment]
changing this issue to a low-priority enhancement

**Comment by gcurtiss** [comment]
as it stands, speed limit is working well enough to where most people aren't complaining. only enhancement would be to fix bypasses (apparently it's pretty easy to bypass the current system). 

@ultra64cmy might revert the warning system because people talk about how easy it is to bypass, probably will go back to how it was originally. still need to look into it more to see how people are bypassing though
